### PR TITLE
use correct manifest list digest for worker build

### PR DESCRIPTION
**Pull request by rcerven**
Signed-off-by: Robert Cerven <rcerven@redhat.com>

### CODE REVIEW

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 1]
```diff
@@ -650,13 +660,159 @@ def set_build_image(self):
...
+            raise RuntimeError("Buildroot image doesn't have tag, which is needed for specified arch")
+
+        # we have build_image with tag, so we can check for manifest list
+        if build_image:
+            registry_name, image = build_image.split('/', 1)
+            registry_name = registry_name.split(':')[0]
+            repo, tag = image.rsplit(':', 1)
+
+            if 'pulp' in registry_name:
+                registry_name += '/pulp/docker'
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Wait, what? That shouldn't be needed. Just use the Docker Registry HTTP API against the registry and port.

Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: yeah I'll change that, as we spoke on IRC

File: atomic_reactor/util.py [hunk 2]
```diff
@@ -916,6 +916,41 @@ def get_manifest_digests(image, registry, insecure=False, dockercfg_path=None,
     return ManifestDigest(**digests)
 
 
+def get_manifest_list(image, registry, insecure=False, dockercfg_path=None):
+    """Return manifest digest for image.
```
Comment (twaugh) [inline atomic_reactor/util.py]: s/digest/list/

File: atomic_reactor/util.py [hunk 3]
```diff
@@ -916,6 +916,41 @@ def get_manifest_digests(image, registry, insecure=False, dockercfg_path=None,
     return ManifestDigest(**digests)
 
 
+def get_manifest_list(image, registry, insecure=False, dockercfg_path=None):
```
Comment (twaugh) [inline atomic_reactor/util.py]: Any possibility this could take a `version` parameter and be used by `get_manifest_digests()`, to reduce duplication?

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 4]
```diff
@@ -650,13 +660,158 @@ def set_build_image(self):
...
-            self.config_kwargs['build_image'] = build_name
+            return build_name
         else:
             raise RuntimeError("Build kind isn't 'DockerImage' but %s" % build_kind)
 
+    def set_build_image(self, platforms):
+        """
+        Overrides build_image for worker, to be same as in orchestrator build
+        """
+        orchestrator_plat = 'x86_64'
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Let's say `platform` instead of `plat` to reduce eye friction when reading the code.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 5]
```diff
@@ -650,13 +660,158 @@ def set_build_image(self):
...
+        status = get_build_json().get("status")
+        build_config_name = None
+        build_image = None
+        imagestream = None
+        need_manifest = not (orchestrator_plat in platforms and len(platforms) == 1)
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: `need_manifest_list`

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 6]
```diff
@@ -650,13 +660,158 @@ def set_build_image(self):
...
+        use_same_image = False
+        current_buildimage = self.get_current_buildimage()
+        self.plat_digests = {}
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: `platform_digests`, and should we initialize this in `__init__`?

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 7]
```diff
@@ -650,13 +660,158 @@ def set_build_image(self):
...
+            try:
+                build_name = build_config['spec']['strategy']['customStrategy']['from']['name']
+                build_kind = build_config['spec']['strategy']['customStrategy']['from']['kind']
+            except KeyError:
+                raise RuntimeError("BuildConfig object is malformed")
+
+            if build_kind == 'DockerImage':
+                build_image = build_name
+            elif build_kind == 'ImageStreamTag':
+                imagestream = build_name
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Can some of this be factored out to avoid repeating it? e.g. `build_image, imagestream = process_from(...)`.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 8]
```diff
@@ -650,13 +660,158 @@ def set_build_image(self):
...
+                build_image = tag_image
+
+        # we don't have build_image and we need it as we want to build on differnt platforms
+        if not build_image and need_manifest:
+            raise RuntimeError("Failed to obtain buidroot image,"
+                               " which is needed for specified arch")
+
+        # we have build_image with sha, but we need tag to check for manifest list,
+        # as we want to build on different platforms
+        if build_image and '@sha256:' in build_image and need_manifest:
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: In the future it's possible the digest will actually be for a manifest list, but right now this is fine.

File: atomic_reactor/util.py (hunk 2 above)
Comment (rcerven) [reply atomic_reactor/util.py]: ok

File: atomic_reactor/util.py (hunk 3 above)
Comment (rcerven) [reply atomic_reactor/util.py]: mmm and how would I use it by get_manifest_digest() ? as that is collection versions: Docker-Content-Digest dict? and for manifest list we don't have it?

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 4 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 5 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 6 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: sure

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 7 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/util.py (hunk 3 above)
Comment (rcerven) [reply atomic_reactor/util.py]: made general get_manifest function, to reduce duplication

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 9]
```diff
@@ -650,13 +673,166 @@ def set_build_image(self):
...
+        else:
+            raise UnknownKindException
+
+        return build_image, imagestream
+
+    def set_build_image(self, platforms):
+        """
+        Overrides build_image for worker, to be same as in orchestrator build
+        """
+        orchestrator_platform = 'x86_64'
```
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: we agreed wit @lcarva that we will assume that, just so we keep backwards compatibility

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 10]
```diff
@@ -650,13 +673,166 @@ def set_build_image(self):
...
+                release = labels['release']
+                version = labels['version']
+                docker_tag = "%s-%s" % version, release
+                build_image = tag_image[:tag_image.find('@sha256')] + ':' + docker_tag
+            else:
+                build_image = tag_image
+
+        # we don't have build_image and we need it as we want to build on differnt platforms
+        if not build_image and need_manifest_list:
+            raise RuntimeError("Failed to obtain buidroot image,"
```
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 11]
```diff
@@ -286,11 +294,24 @@ def __init__(self, tasker, workflow, platforms, build_kwargs,
         self.fs_task_id = self.get_fs_task_id()
         self.release = self.get_release()
 
+        self.plat_des_fallback = []
+        for platform in goarch:
```
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: sure

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lcarva) [reply atomic_reactor/plugins/build_orchestrate_build.py]: Would it be saner to dynamically determine this?
```python
>>> import platform
>>> platform.processor()
'x86_64'
```

[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok using that, and fallback to 'x86_64'

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 12]
```diff
@@ -286,11 +295,24 @@ def __init__(self, tasker, workflow, platforms, build_kwargs,
...
+            'insecure': not verify_ssl,
+            'auth': {'enable': use_auth}
+        }
+
         if worker_build_image:
             self.log.warning('worker_build_image is deprecated')
 
         self.worker_builds = []
         self.namespace = get_build_json().get('metadata', {}).get('namespace', None)
+        self.plat_digests = {}
```
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: we can, but I wanted just to point out that it's keys are platforms

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 13]
```diff
@@ -286,11 +295,24 @@ def __init__(self, tasker, workflow, platforms, build_kwargs,
         self.fs_task_id = self.get_fs_task_id()
         self.release = self.get_release()
 
+        self.plat_des_fallback = []
+        for plat, architecture in goarch.items():
+            plat_dic = {'platform': plat,
+                        'architecture': architecture}
+            self.plat_des_fallback.append(plat_dic)
```
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: we are building fallback for get_platform_descriptors, which is list of dicts

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 14]
```diff
@@ -286,11 +295,24 @@ def __init__(self, tasker, workflow, platforms, build_kwargs,
         self.fs_task_id = self.get_fs_task_id()
         self.release = self.get_release()
 
+        self.plat_des_fallback = []
+        for plat, architecture in goarch.items():
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Needs `if goarch:` in front, or else `for ... in (goarch or {}).items():`

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 15]
```diff
@@ -429,6 +451,11 @@ def get_cluster_info(self, cluster, platform):
         if self.osbs_client_config:
             kwargs['conf_file'] = os.path.join(self.osbs_client_config, 'osbs.conf')
 
+        if platform in self.build_image_digests:
+            kwargs['build_image'] = self.build_image_digests[platform]
+        else:
+            raise RuntimeError("We don't have build_image for platform %s", platform)
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: This error message sounds odd. Maybe: `build_image for platform '%s' not available`

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 16]
```diff
@@ -650,13 +674,167 @@ def set_build_image(self):
...
+            raise UnknownKindException
+
+        return build_image, imagestream
+
+    def set_build_image(self, platforms):
+        """
+        Overrides build_image for worker, to be same as in orchestrator build
+        """
+        current_platform = platform.processor()
+        orchestrator_platform = current_platform if current_platform else 'x86_64'
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Why not simply `orchestrator_platform = platform.processor() or 'x86_64'` ?

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 17]
```diff
@@ -650,13 +674,167 @@ def set_build_image(self):
...
+        status = get_build_json().get("status")
+        build_config_name = None
+        build_image = None
+        imagestream = None
+        need_manifest_list = not (orchestrator_platform in platforms and len(platforms) == 1)
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Simplified alternative: `need_manifest_list = platforms != [orchestrator_platform]`

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 18]
```diff
@@ -650,13 +674,167 @@ def set_build_image(self):
...
+
+        # BuildConfig exists
+        if 'config' in status:
+            config = status['config']
+            if config['kind'] == 'BuildConfig':
+                build_config_name = config['name']
+            else:
+                raise RuntimeError("Build config type isn't BuildConfig : %s" % config['kind'])
+
+        # get image build from build metadata, which is set for direct builds
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Let's clarify that this is "set by osbs-client" to avoid confusion with OpenShift behavior.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 19]
```diff
@@ -650,13 +674,167 @@ def set_build_image(self):
...
+
+        return build_image, imagestream
+
+    def set_build_image(self, platforms):
+        """
+        Overrides build_image for worker, to be same as in orchestrator build
+        """
+        current_platform = platform.processor()
+        orchestrator_platform = current_platform if current_platform else 'x86_64'
+        status = get_build_json().get("status")
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: If this is `None`, `if 'config' in status:` below will blow up.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 20]
```diff
@@ -650,13 +674,167 @@ def set_build_image(self):
...
+
+            if 'from' in metadata['annotations']:
+                scratch_from = json.loads(metadata['annotations']['from'])
+
+                try:
+                    build_image, imagestream = self.process_image_from(scratch_from)
+                except UnknownKindException:
+                    raise RuntimeError("Build annotation has unknown 'kind' %s" %
+                                       scratch_from['kind'])
+            elif not need_manifest_list:
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Should this be the first check this function does? If it doesn't need manifest list, just use the same image and bail out?

This will also ensure that this plugin does not always require an OpenShift session, which is really important for backwards compatibility.

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 19 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: well not sure if it is possible for status of build to be None, but ok will make it default {}

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 15 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 16 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: sure

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 18 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 20 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: well not really, IF there is manifest list used, I'd like to actually verify that we are using correct digest from it

Comment (lcarva) [reply atomic_reactor/plugins/build_orchestrate_build.py]: @rcerven pointed out that the story for this (OSBS-5221) actually calls out for always making this check. I think this was an oversight. @twaugh, think we should only bother with this new behavior if manifest list is actually required (at least one worker build uses different arch than orchestrator build) ?

I'm concerned in making the new plugin parameters for building openshift session required for all cases.

Comment (twaugh) [reply atomic_reactor/plugins/build_orchestrate_build.py]: @lcarva yes, I agree. If we do not need the manifest list (because there is a single worker, for the same architecture as the orchestrator), let's skip the additional processing.

Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ack

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 21]
```diff
@@ -651,13 +675,158 @@ def set_build_image(self):
...
+            if config['kind'] == 'BuildConfig':
+                build_config_name = config['name']
+            else:
+                raise RuntimeError("Build config type isn't BuildConfig : %s" % config['kind'])
+
+        # get image build from build metadata, which is set for direct builds
+        # this is explicitly set by osbs-client, it isn't default OpenShift behaviour
+        else:
+            metadata = get_build_json().get("metadata")
+            if 'from' in metadata['annotations']:
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: 'annotations' is not a required property of v1.ObjectMeta, so maybe:
```
annotations = get_build_json().get('metadata', {}).get('annotations', {})
if 'from' in annotations:
```

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 22]
```diff
@@ -651,13 +675,158 @@ def set_build_image(self):
...
+                tag = osbs.get_image_stream_tag(imagestream).json()
+            except OsbsException:
+                raise RuntimeError("ImageStreamTag not found %s" % imagestream)
+
+            try:
+                tag_image = tag['image']['dockerImageReference']
+            except KeyError:
+                raise RuntimeError("ImageStreamTag is malformed %s" % imagestream)
+
+            if '@sha256:' in tag_image:
```
Comment (twaugh) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Minor (no action): 'sha256:' is part of the digest and in the future a different algorithm may be used. However, let's hope ImageStreams gain manifest list support before then!

File: tests/plugins/test_orchestrate_build.py [hunk 23]
```diff
@@ -11,7 +11,7 @@
 from atomic_reactor.core import DockerTasker
 from atomic_reactor.inner import DockerBuildWorkflow
 from atomic_reactor.plugin import BuildCanceledException, PluginFailedException
-from atomic_reactor.plugin import BuildStepPluginsRunner
+from atomic_reactor.plugin import BuildStepPluginsRunner, PluginsRunner, PreBuildPluginsRunner
```
[the remark this answers is not in the document]
Comment (twaugh) [reply tests/plugins/test_orchestrate_build.py]: Seems like sticker is right.

[the remark this answers is not in the document]
Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: yeah tests are still in progress

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 21 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: tests/plugins/test_orchestrate_build.py [hunk 24]
```diff
@@ -351,6 +369,8 @@ def test_orchestrate_build(tmpdir, caplog, config_kwargs,
                     namespace = {name}_namespace
                     """.format(name=cluster['name'])))
 
+    goarch = {'x86_64': 'amd64', 'ppc64le': 'ppc64'}
```
Comment (twaugh) [inline tests/plugins/test_orchestrate_build.py]: I think it doesn't matter because this is just test data, but the GOARCH for 'ppc64le' is 'ppc64le' so we wouldn't bother including that key. A likely-looking dict would be `{'x86_64': 'amd64', 'aarch64': 'arm64'}`.

File: tests/plugins/test_orchestrate_build.py [hunk 25]
```diff
@@ -97,8 +107,10 @@ def mock_workflow(tmpdir):
     df = df_parser(df_path)
     setattr(workflow.builder, 'df_path', df.dockerfile_path)
 
-    build = '{"spec": {"strategy": {"customStrategy": {"from": \
-            {"name": "some_image:latest", "kind": "DockerImage"}}}}}'
+    build = '{"spec": {"strategy": {"customStrategy": {"from":' + \
+            '{"name": "registry/some_image@sha256:123456", "kind": "DockerImage"}}}},' + \
+            '"metadata": {"annotations": {"from": "{\\"name\\": \\"registry' + \
+            '/some_image:latest\\", \\"kind\\": \\"DockerImage\\"}"}}}'
```
Comment (twaugh) [inline tests/plugins/test_orchestrate_build.py]: There will come a point where it might be easier to just use `json.dumps()` on a dict to get these strings.

File: tests/plugins/test_orchestrate_build.py (hunk 24 above)
Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: well even if it is same it has to be in goarch, or we will need to change logic how we use it, that is used in group_manifest plugin as well

Comment (twaugh) [reply tests/plugins/test_orchestrate_build.py]: No, definitely not:
https://github.com/projectatomic/atomic-reactor/blob/master/atomic_reactor/plugins/post_group_manifests.py#L236

This is particularly important, because we do not have a platform descriptor for ppc64le because it doesn't need one.

File: tests/plugins/test_orchestrate_build.py [hunk 26]
```diff
@@ -1138,48 +1182,144 @@ def test_orchestrate_override_build_kwarg(tmpdir, overrides):
...
+
+    ('{"spec": {"strategy": {"customStrategy": {"from": \
+     {"name": "osbs-buildroot:latest", "kind": "DockerImage"}}}}, \
+     "status": {"config": {"kind": "BuildConfig", "name": "build config"}}}',
+     "Buildroot image doesn't have tag, which is needed for specified arch",
+     True, {"spec": {"strategy": {"customStrategy": {"from": {"kind": "ImageStreamTag",
+                                                              "name": "ims"}}}}},
+     True, {"image": {"dockerImageReference": "some@sha256:12345",
+                      "dockerImageMetadata": {
+                          "Config": {"Labels": {"release": "@sha256:123",
```
Comment (twaugh) [inline tests/plugins/test_orchestrate_build.py]: The 'release' label won't look like that though, more like '4'.

File: tests/plugins/test_orchestrate_build.py (hunk 24 above)
Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: ok will change how to use platform_descriptors and default to platform if not there

File: tests/plugins/test_orchestrate_build.py (hunk 26 above)
Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: yes I know, but I need scenario where I have @256 and this was way how to get it

Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: I'll change it, to get it from annotations

File: atomic_reactor/plugins/post_group_manifests.py [hunk 27]
```diff
@@ -65,9 +65,9 @@ def __init__(self, tasker, workflow, registries=None, group=True, goarch=None):
         self.group = get_group_manifests(self.workflow, group)
 
         plat_des_fallback = []
-        for platform in goarch:
+        for platform, architecture in (goarch or {}).items():
             plat_dic = {'platform': platform,
-                        'architecture': goarch[platform]}
+                        'architecture': architecture}
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/post_group_manifests.py]: what's the reason for this change? if you want to handle the case where goarch=None, couldn't this change just as easily be 'for platform in goarch or {}:'
I don't really object, I'm just curious about why one form is preferred over the other.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 28]
```diff
@@ -651,13 +675,161 @@ def set_build_image(self):
...
+            self.build_image_digests[orchestrator_platform] = current_buildimage
+            return
+
+        # BuildConfig exists
+        if 'config' in status:
+            config = status['config']
+            if config['kind'] == 'BuildConfig':
+                build_config_name = config['name']
+            else:
+                raise RuntimeError("Build config type isn't BuildConfig : %s" % config['kind'])
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/build_orchestrate_build.py]: since set_build_image is still a huge function and status isn't used for anything other than getting the config, I think this can safely be turned into a utility function, ie:
`
try:
    build_config_name = self.get_build_config_name()
except KeyError:
   the current else block
`

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 29]
```diff
@@ -651,13 +675,161 @@ def set_build_image(self):
...
+                try:
+                    build_image, imagestream = self.process_image_from(scratch_from)
+                except UnknownKindException:
+                    raise RuntimeError("Build annotation has unknown 'kind' %s" %
+                                       scratch_from['kind'])
+            else:
+                raise RuntimeError("Build wasn't created from BuildConfig and neither"
+                                   " has 'from' annotation, which is needed for specified arch")
+
+        osbs = get_openshift_session(self.workflow, self.openshift_fallback)
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/build_orchestrate_build.py]: can this be moved up to be above the check for config in status? it doesn't look like it depends on the results of the if else block.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 30]
```diff
@@ -651,13 +675,161 @@ def set_build_image(self):
...
+
+            try:
+                build_from = build_config['spec']['strategy']['customStrategy']['from']
+            except KeyError:
+                raise RuntimeError("BuildConfig object is malformed")
+
+            try:
+                build_image, imagestream = self.process_image_from(build_from)
+            except UnknownKindException:
+                raise RuntimeError("BuildConfig object has unknown 'kind' %s" % build_from['kind'])
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Since you only get a build_config_name if config is in status, this block could be moved up to be part of the if config in status block, and then moved to the function I proposed earlier.

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 31]
```diff
@@ -651,13 +675,161 @@ def set_build_image(self):
...
+            else:
+                build_image = tag_image
+
+        # we have build_image with sha, but we need tag to check for manifest list,
+        # as we want to build on different platforms
+        if build_image and '@sha256:' in build_image:
+            raise RuntimeError("Buildroot image doesn't have tag,"
+                               " which is needed for specified arch")
+
+        # we have build_image with tag, so we can check for manifest list
```
Comment (mlangsdorf) [inline atomic_reactor/plugins/build_orchestrate_build.py]: is it worthwhile splitting this entire block off into its own function? it's clearly very self-contained.

File: tests/plugins/test_orchestrate_build.py (hunk 25 above)
Comment (mlangsdorf) [reply tests/plugins/test_orchestrate_build.py]: I think we've past that point.

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 28 above)
Comment (lcarva) [reply atomic_reactor/plugins/build_orchestrate_build.py]: I'm also not sure if we actually need to perform these checks. If 'config' is set to something other than `BuildConfig` object there's something seriously wrong in OpenShift. 

But anyways, having a utility method that encapsulates this would help in reducing the size of this method. I'd even go further, and just let `get_build_config_name` that Mark suggested raise the RuntimeError with the helpful message. In other words, this whole block turns into:
```python
build_config_name = self.get_build_config_name()  # also fetches the status
```

`get_build_config_name` returns `None` if `config not in status`; or the actual build config name; or raises RuntimeError in case of data error.

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 29 above)
Comment (lcarva) [reply atomic_reactor/plugins/build_orchestrate_build.py]: Could turn into a property so it's only initialized when needed, similar to how `resolve_composes` plugin handles `odcs_client` and `koji_client`.

File: atomic_reactor/plugins/post_group_manifests.py (hunk 27 above)
Comment (rcerven) [reply atomic_reactor/plugins/post_group_manifests.py]: the same change was done in orchestrator, as Tim wanted to use items() since we are using both key&value

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 29 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: at this point we certainly need that session, in other cases it would raise and we wouldn't be at this point

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 30 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: mmm yeah

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 31 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: yeah

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 29 above)
Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok will make it property

File: atomic_reactor/plugins/build_orchestrate_build.py (hunk 30 above)
Comment (mlangsdorf) [reply atomic_reactor/plugins/build_orchestrate_build.py]: I think you can tighten that up more:
build_image, imagestream = get_image_info_from_config()
has the current get_build_config_name and all of the code in the current if build_config_name: block

```
-       if build_config_name:
-       else:
+      if not (build_image or imagetream):
```

File: atomic_reactor/plugins/build_orchestrate_build.py [hunk 32]
```diff
@@ -651,13 +683,169 @@ def set_build_image(self):
...
+        repo, tag = image.rsplit(':', 1)
+
+        registry = ImageName(registry=registry_name, repo=repo, tag=tag)
+        manifest_list = get_manifest_list(registry, registry_name, insecure=True)
+
+        # we don't have manifest list, but we want to build on different platforms
+        if not manifest_list:
+            raise RuntimeError("Buildroot image tag isn't manifest list,"
+                               " which is needed for specified arch")
+        else:
```
Comment (lcarva) [inline atomic_reactor/plugins/build_orchestrate_build.py]: Unnecessary `else`.

Comment (rcerven) [reply atomic_reactor/plugins/build_orchestrate_build.py]: ok

File: tests/plugins/test_orchestrate_build.py (hunk 24 above)
Comment (lcarva) [reply tests/plugins/test_orchestrate_build.py]: Can we also remove `ppc64le` item to ensure the code is not requiring it?

File: tests/plugins/test_orchestrate_build.py [hunk 33]
```diff
@@ -448,7 +488,8 @@ def mock_wait_for_build_to_finish(build_name):
                 'build_kwargs': make_worker_build_kwargs(),
                 'osbs_client_config': str(tmpdir),
                 'max_cluster_fails': 2,
-                'unreachable_cluster_retry_delay': .1
+                'unreachable_cluster_retry_delay': .1,
+                'goarch': {'x86_64': 'amd64', 'ppc64le': 'ppc64'},
```
Comment (lcarva) [inline tests/plugins/test_orchestrate_build.py]: Can we also remove `ppc64le` item to ensure the code is not requiring it? Same for other cases.

File: tests/plugins/test_orchestrate_build.py (hunk 24 above)
Comment (rcerven) [reply tests/plugins/test_orchestrate_build.py]: ok

**Comment by lcarva** [comment]
Updated release notes draft:
> - use correct manifest list digest for worker build image
